src/state.rs

```diff
@@ -341,9 +370,56 @@ impl DemandProxyState {
                     .on_demand_dns_cache_misses
                     .get_or_create(&labels)
                     .inc();
-                // TODO: optimize so that if multiple requests to the same hostname come in at the same time,
+                // optimize so that if multiple requests to the same hostname come in at the same time,
                 // we don't start more than one background on-demand DNS task
-                Self::resolve_on_demand_dns(self.to_owned(), workload).await;
+                if state
+                    .get_cached_resolve_dns_for_hostname(&hostname)
```

Comment:
we should directly check state.get_ips_for_hostname(&hostname).is_none here

and then could remove the the resolved hostnames from `crd_hashmap` to save memory

Reply:
crd_hashmap has a potential mem leak

Reply:
Hi @hzxuzhonghu, the crd hashmap need to store the information for hostname which need to be resolved, so what's kind of case would leak the mem?